client/src/_session_layer.rs

```diff
@@ -0,0 +1,1536 @@
+use futures::{FutureExt, SinkExt, TryFutureExt};
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::convert::{TryFrom, TryInto};
+use std::net::SocketAddr;
+use std::sync::{Arc, Mutex, Weak};
+use tokio::sync::RwLock;
+
+use crate::_client::{ClientConfig, Forwarded};
+use crate::_direct_session::{DirectSession, NodeEntry};
+use crate::_dispatch::{dispatch, Handler};
+use crate::_encryption::Encryption;
+use crate::_error::{ProtocolError, ResultExt, SessionError, SessionInitError, SessionResult};
+use crate::_expire::track_sessions_expiration;
+use crate::_raw_session::{RawSession, SessionType};
+use crate::_routing_session::{NodeRouting, RoutingSender};
+use crate::_session_protocol::SessionProtocol;
+use crate::_session_registry::{Registry, SessionLock, SessionPermit, Validity};
+use crate::_session_state::{RelayedState, SessionState};
+use crate::_session_traits::{SessionDeregistration, SessionRegistration};
+use crate::_transport_layer::ForwardReceiver;
+
+use ya_relay_core::identity::Identity;
+use ya_relay_core::server_session::{Endpoint, NodeInfo, SessionId, TransportType};
+use ya_relay_core::udp_stream::{udp_bind, OutStream};
+use ya_relay_core::utils::spawn_local_abortable;
+use ya_relay_core::{challenge, NodeId};
+use ya_relay_proto::codec::PacketKind;
+use ya_relay_proto::proto;
+use ya_relay_proto::proto::control::disconnected::By;
+use ya_relay_proto::proto::control::ReverseConnection;
+use ya_relay_proto::proto::{Forward, RequestId, SlotId, FORWARD_SLOT_ID};
+use ya_relay_stack::Channel;
+
+type ReqFingerprint = (Vec<u8>, u64);
+
+#[derive(Copy, Clone, Display, PartialEq, Eq)]
+pub enum ConnectionMethod {
+    Direct,
+    Reverse,
+    Relay,
+    // Here should be NAT punching type(s) defined in the future
+}
+
+/// Responsible for establishing/receiving connections from other Nodes.
+/// Hides from upper layers the decisions, how to route packets to desired location
+#[derive(Clone)]
+pub struct SessionLayer {
+    pub config: Arc<ClientConfig>,
+    /// Could be just `Option<OutStream>` but then we are not able to drop all senders
+    /// when we are copying `SessionLayer`, what prevents clean shutdown.
+    sink: Arc<Mutex<Option<OutStream>>>,
+
+    pub(crate) state: Arc<RwLock<SessionLayerState>>,
+
+    pub(crate) registry: Registry,
+    ingress_channel: Channel<Forwarded>,
+
+    // TODO: Could be per `Session`.
+    processed_requests: Arc<Mutex<VecDeque<ReqFingerprint>>>,
+}
+
+#[derive(Default)]
+pub struct SessionLayerState {
+    /// If address is None after registering endpoints on Server, that means
+    /// we don't have public IP.
+    public_addr: Option<SocketAddr>,
+    /// Equals to `None` when not listening
+    pub(crate) bind_addr: Option<SocketAddr>,
+
+    /// Maps contains mapping for default and secondary identities.
+    pub nodes: HashMap<NodeId, Arc<NodeRouting>>,
+    pub p2p_sessions: HashMap<SocketAddr, Arc<DirectSession>>,
+    pub p2p_nodes: HashMap<NodeId, Arc<DirectSession>>,
```

Comment:
For instance it is not clear that in case:
- relyed connection: there is entry in `nodes` and it points to the `p2p_sessions` & `p2p_nodes` entry?
- p2p connection: is there new entry in `nodes` or not?